### ISSUE TITLE
Suggestion: throw an exception if the library is used before initialization

### ISSUE THREAD

**Issue by MaksymB** [log4cplus/log4cplus#276]
If the library states that it cannot be used before it is initialized at the top of main, it would be very helpful to tell the user (writing to console and/or throwing an exception) if this precondition is not met. Current implementation tries to initialize the library anyway, but not always succeeds, sometimes these attempts are resulting in a dead lock.
If this is a Windows-specific problem, would be nice at least to be able to switch on such a "cautious" mode via build configuration parameters.

What exactly I suggest is
- to remove initialization from `DLL_PROCESS_ATTACH` section in `thread_callback` (global-init.cxx);
- to change `get_dc(alloc)`'s default argument to `false` and it's body to something like this:
```
static DefaultContext * get_dc (bool alloc = false)
{
    if (LOG4CPLUS_UNLIKELY(!default_context))
    {
        if (alloc)
        {
            alloc_dc();
        }
        else
        {
            std::cout << "log4cplus is not initialized" << std::endl;
            throw std::logic_error("log4cplus is not initialized");
        }
    }
    return default_context;
}
```

**Comment by MaksymB** [comment]
@wilx could you, please, comment what are your plans concerning this issue?

**Comment by wilx** [comment]
TBH, I am not sure. It is convenient to have the library initialize itself in most situations.

**Comment by MaksymB** [comment]
In most cases - yes, but when it fails it's a big issue. I will explain what bothers me on daily basis. I release a few DLLs with log4cplus-based logging, which are used by many applications. It's not always obvious that the application needs to initialize logger even if it's not directly used. And everything works fine in most cases but sometimes the applications hang on start or exit. IMHO, this faithless seldom hang is much worse than honest crash (via unhandled exception) on the first launch on developer's machine.

**Comment by wilx** [comment]
Merged in #404 and #405.
